### ISSUE TITLE
unsafe_load for non-bitstype pointers is more unsafe than necessary

### ISSUE THREAD

**Issue by chethega** [JuliaLang/julia#30276]
```
julia> function foo()
       A=[([],), ([],)];
       x=unsafe_load(pointer(A));
       A[2]=x;
       p1 = unsafe_load(reinterpret(Ptr{UInt64},pointer(A)), 1)
       p2 = unsafe_load(reinterpret(Ptr{UInt64},pointer(A)), 2)
       p3 = unsafe_load(reinterpret(Ptr{UInt64},p2))
       @show p1,p2,p3
       return A
       end

julia> r=foo();
(p1, p2, p3) = (0x00007f2f0543f8d0, 0x00007f2f05638110, 0x00007f2f0543f8d0)

julia> r
2-element Array{Tuple{Array{Any,1}},1}:

signal (11): Segmentation fault
```
I am not entirely sure whether this is supposed to be working. If not, then maybe `unsafe_load` could error out earlier. 

If this is supposed to be working, we see that someone got confused about types / numbers of indirection: `A[2]` points to something that points to `A[1]`, instead of pointing to something that points to `A[1][1]` (like e.g. `A[1]`). Also reproduces in the REPL (ie without compilation). Tested on 1.0.2.

**Comment by KristofferC** [comment]
DId this come up in any real code for which this is a MWE of?

**Comment by chethega** [comment]
No; when working with constructions that are new to me (like above) I tend to run some tests before deciding whether to structure my code around them. I think that the correct way is to `ccall` something to get an object from a pointer. But since the above doesn't throw an error (honestly what I expected) and is the obvious syntax, well, the result cannot be intended.

**Comment by yuyichao** [comment]
Yes `unsafe_load` is supposed to work, but not supposed to work with your input. The correct input is a pointer to the type, not a pointer to a slot for the pointer to that type.

```julia
julia> A = [([],), ([],)]
2-element Array{Tuple{Array{Any,1}},1}:
 ([],)
 ([],)

julia> B = Base.cconvert(Ref{Tuple{Vector{Any}}}, A[1])
Base.RefValue{Tuple{Array{Any,1}}}((Any[],))

julia> Base.unsafe_convert(Ref{Tuple{Vector{Any}}}, B)
Ptr{Tuple{Array{Any,1}}} @0x00007fd11de85520

julia> unsafe_load(Base.unsafe_convert(Ref{Tuple{Vector{Any}}}, B))
(Any[],)
```

The fundamental issue if that `Ptr` of non bitstype is ambiguious. In the current state, it makes sense for both `pointer(A)` and `Base.unsafe_convert(Ref{Tuple{Vector{Any}}}, B)` to return such a type and this is not fixable without a breaking change that changes this.

**Comment by yuyichao** [comment]
In another word, the correct fix is to make `pointer(A)` returns `Ptr{Ptr{Tuple{Vector{Any}}}}` instead. That will be a breaking change.

**Comment by chethega** [comment]
Changing the type of `pointer(A)` would be a nice fix.

How breaking would that change be? That is, there are not a lot of useful things people can do with the current pointer type. In order to avoid future breakage if any kind of inline storage comes, `Ptr{Nothing}` could also be a sane return type (effectively `void*`).

**Comment by chethega** [comment]
`Ptr{Nothing}===Ptr{Cvoid}` is the return-type of `pointer_from_objref`. But `Ptr{Ptr{Nothing}}` or `Ptr{Ptr{Any}}` or the correct type (`Ptr{Ptr{T}}`) might be better for non-inline stored arrays. For bitsunions, I agree with @mbauman that there is no sensible type, so it wants to be untyped (whether that means `Ptr{Nothing}`, `Ptr{Cvoid}` or `Ptr{Union{}}`).

**Comment by FedericoStra** [comment]
Is there interest in fixing this? I could prepare a PR based on the code presented [here](https://github.com/JuliaLang/julia/issues/38848#issue-763974566), but I want to make sure that this kind of fix would be accepted and not waste time writing something that gets rejected a priori.
